File: frontend/src/types/protocol.ts
export interface MouseEvent {
  type: 'mouse_event'
  x: number
  y: number
  button: 'left' | 'right' | 'middle' | 'none'
  action: 'move' | 'down' | 'up' | 'scroll'
  scrollX?: number
  scrollY?: number
}

export interface KeyboardEvent {
  type: 'keyboard_event'
  key: string
  code: string
  action: 'down' | 'up'
  modifiers: {
    ctrl: boolean
    alt: boolean
    shift: boolean
    meta: boolean
  }
}

export interface ScreenInfo {
  type: 'screen_info'
  width: number
  height: number
  monitorIndex: number
  encoding: 'jpeg' | 'h264'
}

export interface MonitorList {
  type: 'monitor_list'
  monitors: { index: number; name: string; width: number; height: number; primary: boolean }[]
}

export interface MonitorSwitch {
  type: 'monitor_switch'
  index: number
}

export interface ConfigUpdate {
  type: 'config_update'
  fps?: number
  quality?: number
}

export interface ControlRequest {
  type: 'control_request'
}

export interface ControlGrant {
  type: 'control_grant'
  granted: boolean
  controller?: string
}

export interface ClipboardSync {
  type: 'clipboard_sync'
  text: string
}

export interface Stats {
  type: 'stats'
  bandwidth: number
  fps: number
  encoding: string
}

export interface SpecialKey {
  type: 'special_key'
  key: 'ctrl_alt_del' | 'win' | 'print_screen'
}

export interface ChatMessage {
  type: 'chat_message'
  id: string
  sender: string
  text: string
  timestamp: number
}

export interface Ping {
  type: 'ping'
  timestamp: number
}

export interface Pong {
  type: 'pong'
  timestamp: number
}

export type WSMessage =
  | MouseEvent
  | KeyboardEvent
  | ScreenInfo
  | MonitorList
  | MonitorSwitch
  | ConfigUpdate
  | ControlRequest
  | ControlGrant
  | ClipboardSync
  | Stats
  | SpecialKey
  | ChatMessage
  | Ping
  | Pong

export interface FileEntry {
  name: string
  path: string
  isDir: boolean
  size: number
  modTime: string
  mode: string
}

export interface SysInfo {
  hostname: string
  os: string
  arch: string
  cpuModel: string
  cpuCores: number
  cpuUsage: number
  memTotal: number
  memUsed: number
  gpu: string
  network: { name: string; ip: string }[]
}

export interface ProxyRule {
  id: string
  pathPrefix: string
  target: string
}

export interface AuthResponse {
  token: string
  username: string
}

File: backend/ws/protocol.go
package ws

type MessageType string

const (
	MsgMouseEvent     MessageType = "mouse_event"
	MsgKeyboardEvent  MessageType = "keyboard_event"
	MsgScreenInfo     MessageType = "screen_info"
	MsgMonitorList    MessageType = "monitor_list"
	MsgMonitorSwitch  MessageType = "monitor_switch"
	MsgConfigUpdate   MessageType = "config_update"
	MsgControlRequest MessageType = "control_request"
	MsgControlGrant   MessageType = "control_grant"
	MsgClipboardSync  MessageType = "clipboard_sync"
	MsgStats          MessageType = "stats"
	MsgSpecialKey     MessageType = "special_key"
	MsgChatMessage    MessageType = "chat_message"
	MsgPing           MessageType = "ping"
	MsgPong           MessageType = "pong"
)

type BaseMessage struct {
	Type MessageType `json:"type"`
}

type MouseEventMsg struct {
	Type    MessageType `json:"type"`
	X       float64     `json:"x"`
	Y       float64     `json:"y"`
	Button  string      `json:"button"`
	Action  string      `json:"action"`
	ScrollX float64     `json:"scrollX,omitempty"`
	ScrollY float64     `json:"scrollY,omitempty"`
}

type KeyboardEventMsg struct {
	Type      MessageType       `json:"type"`
	Key       string            `json:"key"`
	Code      string            `json:"code"`
	Action    string            `json:"action"`
	Modifiers KeyboardModifiers `json:"modifiers"`
}

type KeyboardModifiers struct {
	Ctrl  bool `json:"ctrl"`
	Alt   bool `json:"alt"`
	Shift bool `json:"shift"`
	Meta  bool `json:"meta"`
}

type ScreenInfoMsg struct {
	Type         MessageType `json:"type"`
	Width        int         `json:"width"`
	Height       int         `json:"height"`
	MonitorIndex int         `json:"monitorIndex"`
	Encoding     string      `json:"encoding"`
}

type MonitorInfo struct {
	Index   int    `json:"index"`
	Name    string `json:"name"`
	Width   int    `json:"width"`
	Height  int    `json:"height"`
	Primary bool   `json:"primary"`
}

type MonitorListMsg struct {
	Type     MessageType   `json:"type"`
	Monitors []MonitorInfo `json:"monitors"`
}

type MonitorSwitchMsg struct {
	Type  MessageType `json:"type"`
	Index int         `json:"index"`
}

type ConfigUpdateMsg struct {
	Type    MessageType `json:"type"`
	FPS     *int        `json:"fps,omitempty"`
	Quality *int        `json:"quality,omitempty"`
}

type ControlRequestMsg struct {
	Type MessageType `json:"type"`
}

type ControlGrantMsg struct {
	Type       MessageType `json:"type"`
	Granted    bool        `json:"granted"`
	Controller string      `json:"controller,omitempty"`
}

type ClipboardSyncMsg struct {
	Type MessageType `json:"type"`
	Text string      `json:"text"`
}

type StatsMsg struct {
	Type      MessageType `json:"type"`
	Bandwidth float64     `json:"bandwidth"`
	FPS       float64     `json:"fps"`
	Encoding  string      `json:"encoding"`
}

type SpecialKeyMsg struct {
	Type MessageType `json:"type"`
	Key  string      `json:"key"`
}

type ChatMessageMsg struct {
	Type      MessageType `json:"type"`
	ID        string      `json:"id"`
	Sender    string      `json:"sender"`
	Text      string      `json:"text"`
	Timestamp int64       `json:"timestamp"`
}

type PingMsg struct {
	Type      MessageType `json:"type"`
	Timestamp int64       `json:"timestamp"`
}

type PongMsg struct {
	Type      MessageType `json:"type"`
	Timestamp int64       `json:"timestamp"`
}
